src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp

```diff
@@ -177,10 +177,13 @@ void ShenandoahControlThread::run_service() {
       // it is a normal completion, or the abort.
       heap->free_set()->log_status_under_lock();
 
-      // Notify Universe about new heap usage. This has implications for
-      // global soft refs policy, and we better report it every time heap
-      // usage goes down.
-      heap->update_capacity_and_used_at_gc();
+      {
+        // Notify Universe about new heap usage. This has implications for
+        // global soft refs policy, and we better report it every time heap
+        // usage goes down.
+        ShenandoahHeapLocker locker(heap->lock());
+        heap->update_capacity_and_used_at_gc();
+      }
```

Comment:
If free set logging is enabled this does a lock/unlock for logging and then another lock/unlock for updating capacity.

But I guess it's unavoidable, and it likely won't be the case that we have free set logging enabled in performance critical production situations, so not worth wasting too much sleep over.

Reviewed!

Reply:
@earthling-amzn : I'd suggest leaving a comment either in the ticket or in the PR (or here in the code?) stating the race that we are vulnerable to if we don't hold the lock, namely the potential skew between used & capacity, and why it's important not to have that skew for consumers of these fields. I imagine that we can demonstrate the issue with a targeted 
 regression test.

Reply:
Yeah, I think coupling the lock with the logging and `update_capacity_and_used` is what led us into trouble to begin with.